src/typeInference/annotateAST.ts

```diff
@@ -13,9 +14,331 @@ import { annotateLogicalOperator } from './annotateLogicalOperator';
+		const plusVal = this.visit(astHelper.getFirstChild(ast, 'operand')[1] as IAST, options);
+		return annotateUnaryPlus(ast, plusVal);
+	}
+
+	visitBinaryOp(ast: IAST, options: StaticContext): SequenceType {
+		const left = this.visit(astHelper.getFirstChild(ast, 'firstOperand')[1] as IAST, options);
+		const right = this.visit(astHelper.getFirstChild(ast, 'secondOperand')[1] as IAST, options);
+		return annotateBinOp(ast, left, right, ast[0]);
+	}
+
+	visitSequenceExpr(ast: IAST, options: StaticContext): SequenceType {
+		const children = astHelper.getChildren(ast, '*');
+		children.map((arg) => this.visit(arg, options));
+		return annotateSequenceOperator(ast, children.length);
+	}
+
+	visitUnionOp(ast: IAST, options: StaticContext): SequenceType {
+		const l = this.visit(astHelper.getFirstChild(ast, 'firstOperand')[1] as IAST, options);
+		const r = this.visit(astHelper.getFirstChild(ast, 'secondOperand')[1] as IAST, options);
+		return annotateSetOperator(ast, l, r);
+	}
+
+	visitIntersectExcept(ast: IAST, options: StaticContext): SequenceType {
+		const l = this.visit(astHelper.getFirstChild(ast, 'firstOperand')[1] as IAST, options);
+		const r = this.visit(astHelper.getFirstChild(ast, 'secondOperand')[1] as IAST, options);
+		return annotateSetOperator(ast, l, r);
+	}
+
+	visitStringConcatenateOp(ast: IAST, options: StaticContext): SequenceType {
+		const stringSequenceType = {
+			type: ValueType.XSSTRING,
+			mult: SequenceMultiplicity.EXACTLY_ONE,
+		};
+
+		astHelper.insertAttribute(ast, 'type', stringSequenceType);
+		return stringSequenceType;
+	}
+
+	visitRangeSequenceExpr(ast: IAST, options: StaticContext): SequenceType {
+		this.visit(astHelper.getFirstChild(ast, 'startExpr')[1] as IAST, options);
+		this.visit(astHelper.getFirstChild(ast, 'endExpr')[1] as IAST, options);
+		return annotateRangeSequenceOperator(ast);
+	}
+
+	visitCompare(ast: IAST, options: StaticContext): SequenceType {
+		switch (ast[0]) {
+			case 'equalOp':
+			case 'notEqualOp':
+			case 'lessThanOrEqualOp':
+			case 'lessThanOp':
+			case 'greaterThanOrEqualOp':
+			case 'greaterThanOp': {
+				this.visit(astHelper.getFirstChild(ast, 'firstOperand')[1] as IAST, options);
+				this.visit(astHelper.getFirstChild(ast, 'secondOperand')[1] as IAST, options);
+				return annotateGeneralCompare(ast);
+			}
+			case 'eqOp':
+			case 'neOp':
+			case 'ltOp':
+			case 'leOp':
+			case 'gtOp':
+			case 'geOp': {
+				this.visit(astHelper.getFirstChild(ast, 'firstOperand')[1] as IAST, options);
+				this.visit(astHelper.getFirstChild(ast, 'secondOperand')[1] as IAST, options);
+				return annotateValueCompare(ast);
+			}
+			case 'nodeBeforeOp':
+			case 'nodeAfterOp': {
+				this.visit(astHelper.getFirstChild(ast, 'firstOperand')[1] as IAST, options);
+				this.visit(astHelper.getFirstChild(ast, 'secondOperand')[1] as IAST, options);
+				return annotateNodeCompare(ast);
+			}
+		}
+	}
+
+	visitPathExpr(ast: IAST, options: StaticContext): SequenceType {
```

Comment:
We probably should actually implement a bunch of these before merging it back into the main branch.